bsp/Src/BSP_ADC.c

```diff
@@ -0,0 +1,163 @@
+#include "BSP_ADC.h"
+
+ADC_HandleTypeDef hadc;
+
+double ADC_RAW_TO_VOLTAGE;  // conversion factor for readings
+
+static SemaphoreHandle_t adc_send_semaphore = NULL;
+static StaticSemaphore_t adc_send_semaphore_buffer;
+
+static QueueHandle_t* adcReadings;
+
+HAL_StatusTypeDef BSP_ADC_Init(ADC_InitTypeDef init, uint8_t bitNum, double Vcc, QueueHandle_t *rxQueue) {
+    
+    ADC_ChannelConfTypeDef sConfig = {0};
+
+    ADC_RAW_TO_VOLTAGE = 1.0 / Vcc * (float) bitNum; 
+
+    adcReadings = rxQueue;
+
+    adc_send_semaphore = xSemaphoreCreateBinaryStatic(&adc_send_semaphore_buffer);
+    xSemaphoreGive(adc_send_semaphore);
+
+    hadc.Init = init; // set the init structure to InitTypeDef
+    if (HAL_ADC_Init(&hadc) != HAL_OK) Error_Handler();
+}
+
+
+HAL_StatusTypeDef BSP_ADC_OneShotRead(uint32_t channel, uint32_t samplingTime, bool blocking) {
+
+    UBaseType_t space = uxQueueSpacesAvailable(adcReadings);    // check for empty space in the queue
+    if (space != 0) {
+        Error_Handler();
+
+        // could xQueueOverwrite last entry instead
+    }
+    
+    TickType_t timeout = blocking ? portMAX_DELAY : 0; // Set timeout to max delay if blocking, 0 if non-blocking
+
+    // take semphr
+    if (xSemaphoreTake(adc_send_semaphore, timeout) != pdTRUE) {
+        Error_Handler();
+    }
+
+    ADC_ChannelConfTypeDef sConfig = {0}; // resets the channel config
+
+    // configure only one channel
+    sConfig.Channel = channel;
+    sConfig.Rank = 1;
+    sConfig.SamplingTime = samplingTime;
+
+    if (HAL_ADC_ConfigChannel(&hadc, &sConfig) != HAL_OK) {
+        Error_Handler();
+    }
+
+    // trigger interrupt to read 
+    HAL_ADC_Start_IT(&hadc);   
+}
+
+void HAL_ADC_ConvCpltCallback(ADC_HandleTypeDef *h) {
+    /*
+    Interrupt triggers this callback when the conversion is complete
+    */
+   BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    if (h == &hadc) {
+        int rawVal = HAL_ADC_GetValue(&hadc);
+        double realVal = (double) rawVal * ADC_RAW_TO_VOLTAGE;
+
+        portENTER_Critical();
+
+        // push value to q
+        xQueueSendFromISR(adcReadings, &realVal, &higherPriorityTaskWoken);
+
+        // give semphr
+        xSemaphoreGive(adc_send_semaphore);
+
+        portEXIT_Critical();
```

Comment:
This doesn't seem like it's a critical section. The QueueSendFromISR already internally enters & exits a critical section, which means these portENTER_Critical/portEXIT_Critical calls aren't required (xQueueSendFromISR is already threadsafe). Furthermore, as explained elsewhere, I don't think this semaphore around the queue send is required either